### PR TITLE
Preserve compressed agent data and compress agent data that is not already compressed

### LINKED ISSUE #71: Don't decompress agent data
When agents send compressed data, the extension should not decompress it, just to compress it again when sending the data to APM Server.

Buffering compressed data is also lighter on memory.

The extension needs to remember the `Content-Encoding` header the agent has sent and use the same when forwarding the data to APM Server.

See also https://github.com/elastic/apm/pull/518#issuecomment-948538762

### PR DESCRIPTION
**Pull request by estolfo**
This PR resolve #71 

The changes include:
- Create a struct for representing the agent data. It has members for the agent data and content encoding of the request
- Change the type of the agent data channel to the `AgentData` struct
- Don't decompress the agent data when it is received by the extension
- When the extension sends the data to the apm server, compress data that is not already compressed and just pass along the content encoding and unchanged data otherwise.
- Remove the `getDecompressedBytesFromRequest` function, as we won't need it anymore
- Remove the tests for the `getDecompressedBytesFromRequest` function

### CODE REVIEW

File: apm-lambda-extension/extension/apm_server.go
Comment (estolfo) [inline apm-lambda-extension/extension/apm_server.go:38]: @felixbarny Should we use gzip to compress the agent data that is not already compressed when sending to the apm server?

Comment (felixbarny) [reply apm-lambda-extension/extension/apm_server.go:38]: GZip is fine but we should compress with the best speed compression level

File: apm-lambda-extension/extension/apm_server.go
Comment (felixbarny) [inline apm-lambda-extension/extension/apm_server.go:37]: Use an `io.Pipe()` to avoid the temporary buffer.

File: apm-lambda-extension/extension/route_handlers.go
Comment (astorm) [inline apm-lambda-extension/extension/route_handlers.go:79]: It looks like `ioutil.ReadAll` doesn't promise success.  Should we be handling the error here instead of suppressing with an `_`? Or is that not needed?

**Review by astorm — COMMENTED** [review]
Looks like a great start and the general shape of things looks correct to me.  Two things to address and (I think) we'll be set 

- I've not done it before but Felix's suggestion of using an `io.Pipe()` to avoid the `compressedBytes` buffer variable is sound

- I have a question about error handling in the new code that reads the agent request body that might lead to a little extra work (or an explanation)

File: apm-lambda-extension/extension/route_handlers.go
Comment (estolfo) [reply apm-lambda-extension/extension/route_handlers.go:79]: I'll ask what the apm server does, as we should probably handle the error similarly to how they would.

File: apm-lambda-extension/extension/apm_server.go
Comment (estolfo) [reply apm-lambda-extension/extension/apm_server.go]: done

**Comment by estolfo** [comment]
@felixbarny If the extension tries to compress non-compressed agent data and fails, what should the extension do? Send the data non-compressed?

File: apm-lambda-extension/extension/route_handlers.go
Comment (estolfo) [reply apm-lambda-extension/extension/route_handlers.go:79]: https://github.com/elastic/apm-server/blob/master/beater/api/intake/handler.go#L72-L100

If the apm server can't read the request, it doesn't log the error but rather writes the error back to the client. I'm not sure we would want to write errors back to the agent, as the agent can't do anything more with the information than the extension can, and we don't want to hold up the lambda function. It seems to me like the best way to handle this would be to log the error and move on. @astorm what do you think? Maybe @felixbarny has some thoughts as well.

**Comment by felixbarny** [comment]
The compression happens while sending data to APM Server. That means if an error happens, some data might have already been sent to the Server. I think the primary source of errors would be network issues rather than issues related to compressing the data. In these cases, I think it's fine to log the error and discard the data.

File: apm-lambda-extension/extension/route_handlers.go
Comment (felixbarny) [reply apm-lambda-extension/extension/route_handlers.go:79]: SGTM. If we find use cases that require sending a response to the agent, we can always add it later.

File: apm-lambda-extension/extension/apm_server.go
Comment (felixbarny) [inline apm-lambda-extension/extension/apm_server.go:46]: [minor suggestion]
IIUC, this function will only be executed once the request has been established and is then streamed (via chunked-encoding) to the server.
If that's correct, the message may be a bit misleading as the more common source of failure is that there's a network issue when streaming the data to APM Server.

```suggestion
				log.Printf("Failed to send compressed data: %v", err)
```

**Review by felixbarny — APPROVED** [review]

File: apm-lambda-extension/extension/route_handlers.go
Comment (astorm) [reply apm-lambda-extension/extension/route_handlers.go:79]: @estolfo logging the error and moving on seem sufficient to me as well.

**Review by astorm — COMMENTED** [review]
Happy to approve once we have logging for the error case on 
```
rawBytes, _ := ioutil.ReadAll(r.Body)
```

File: apm-lambda-extension/extension/apm_server.go
Comment (estolfo) [reply apm-lambda-extension/extension/apm_server.go:46]: We are actually not streaming data yet, but rather sending in batches. At this point in the code, the actual failure is that the data could not be compressed in a go routine. If compression failed in that go routine, we will also get an error that is logged on line 52 (`failed to create a new request when posting to APM server`) and the function will return. So I think the error message here is accurate and should stay as-is.

Comment (felixbarny) [reply apm-lambda-extension/extension/apm_server.go:46]: I think this streams the batches. See also https://medium.com/stupid-gopher-tricks/streaming-data-in-go-without-buffering-3285ddd2a1e5#2342

Comment (estolfo) [reply apm-lambda-extension/extension/apm_server.go:46]: Yes, you're right that the actual send could be chunked by the go internal transport code, I thought you were referring to the way that we send the data as it's received from the agent. But at this point in the code, I'm calling `io.Copy`, which ends up calling [Write](https://pkg.go.dev/compress/gzip#Writer.Write) on the `gzipWriter` before it gets to the network code. So any compression failures would be returned by `io.Copy`. Network errors would be returned by [resp, err := client.Do(req)](https://github.com/elastic/apm-aws-lambda/pull/72/files#diff-6c889602ebc4f41bed697c18ef699798a40df79b104d827330a8bc38be305324R71).

But that actually brings up another point: the error returned from `client.Do(req)` actually wraps the compression error, which means we don't really need to do anything with the error returned from `io.Copy`. In other words, we could removing the logging of the error and just rely on the error returned in [this line](https://github.com/elastic/apm-aws-lambda/pull/72/files#diff-6c889602ebc4f41bed697c18ef699798a40df79b104d827330a8bc38be305324R73)

Comment (felixbarny) [reply apm-lambda-extension/extension/apm_server.go:46]: > I'm calling io.Copy, which ends up calling Write on the gzipWriter before it gets to the network code.

What I'm understanding from the blog that I linked is that `io.Copy` blocks until the `PipeReader` is being read from. This happens when `client.Do(req)` is invoked and the data from the `PipeReader` is written into the HTTP request body via chunked encoding. It's because `io.Copy` blocks that we have to execute it in a concurrently running goroutine. If we didn't do that, we'd never get to the `client.Do(req)` part.

See also the docs for [`io.Pipe`](https://pkg.go.dev/io#Pipe)
> each Write to the PipeWriter blocks until it has satisfied one or more Reads from the PipeReader that fully consume the written data. The data is copied directly from the Write to the corresponding Read (or Reads); there is no internal buffering.

**Pull request merged**
